File: Java/src/_0275_H_Index_II/Solution.java
package _0275_H_Index_II;

/**
 * Date: 2021/2/4 8:50
 * Content:
 * 执行用时：0 ms, 在所有 Java 提交中击败了100.00%的用户
 * 内存消耗：45.5 MB, 在所有 Java 提交中击败了19.18%的用户
 */
public class Solution {

    public int hIndex(int[] c) {

        int n = c.length;
        int l = 0, r = n;
        while (l < r) {
            int mid = l + r + 1 >> 1;
            if (c[n - mid] >= mid) l = mid;
            else r = mid - 1;
        }
        return r;
    }

    public static void main(String[] args) {

        int[] c = {0, 1, 3, 5, 6};
        System.out.println((new Solution()).hIndex(c));  // 3
    }
}

File: Python/_0275_H_Index_II/Solution.py
# coding=utf-8
# Date: 2021/7/12 8:54
from typing import List


# 执行用时：44 ms, 在所有 Python3 提交中击败了62.83%的用户
# 内存消耗：19.1 MB, 在所有 Python3 提交中击败了5.20%的用户
class Solution:
    def hIndex(self, cs: List[int]) -> int:
        n = len(cs)
        l, r = 0, n
        while l < r:
            mid = l + r + 1 >> 1
            if cs[n - mid] >= mid:
                l = mid
            else:
                r = mid - 1
        return r


if __name__ == "__main__":
    print(Solution().hIndex([0, 1, 3, 5, 6]))  # 3
